lib/pbench/test/unit/server/test_requests.py

```diff
@@ -96,7 +96,8 @@ def test_missing_authorization_header(client, caplog, server_config):
         )
         assert response.status_code == 401
         for record in caplog.records:
-            assert record.levelname not in ("ERROR", "CRITICAL")
+            assert record.levelname not in ("DEBUG", "ERROR", "CRITICAL")
+        assert caplog.records[-1].levelname == "WARNING"
```

Comment:
I've always been a bit uncomfortable with tests like this. First, the logging isn't really a part of the API contract, and white box testing is always fragile. The correctness of the API doesn't really hinge on whether we write a log message; and, furthermore, it would be as correct if we issued a `warn` level for the bad header and then an `info` or `debug` for completion of the API, possibly wrapped somewhere implicitly. (That is, I'd be a lot less unhappy with an assertion that we generated at least one `WARNING` level than with an assertion that the _last_ log was `WARNING`.)

The test may be correct as written, now; but it might fail in the future for non-obvious reasons that have nothing to do with whether the API is behaving correctly, and that makes me nervous. It seems this sort of capture is over-testing "just because we can" and we may pay for it later.

Reply:
OK, as a temporary replacement for an audit log record, which I agree _is_ a part of the API, I'll accept this hack. ;-)

Reply:
I agree with Dave's concerns around testing the log contents (particularly against the notion that the test can predict what the last record will be).

I'm concerned about calcifying the contents of the log (Peter posted elsewhere about "Day 1" vs. "Day 2", and I think that the converse argument applies here -- the log should be a "developer artifact", used for diagnosis and support as well as development, and it shouldn't be part of "correct operation".)

As for "too strict", I've worked on a project which used TDD, and it became very hard to refactor the code because the "units" which the unit testing tested were too small, such that virtually any change to the code resulted in having to rewrite the unit tests as well (and, because of the way the unit testing framework worked, it wasn't easy to write anything but whitebox tests).  So, I'd like to avoid that, here:  we need to agree what the interface contract is, and test that, and try not to peek too much under the covers.

Nevertheless, until we have an audit system, the log will have to serve, I suppose.

Reply:
Tracked in #2281.